### PR TITLE
Use watcher that supports paths in update marker

### PR DESCRIPTION
**Pull request by pchila**
## What does this PR do?

This is the last subset of a bigger PR #3950, so CI should be green.

This change checks if the target version of the upgrade is at least 8.13.0 to decide which upgrade watcher to launch. This is needed to ensure that the watcher performs correct cleanup and rollbacks by reading paths in the update marker.

In this PR it's also included a necessary fix to the shutdownCallback since we do some path manipulation to support the legacy structure of component run directory where we replace versions from old to new.

This PR includes also a set of minor fixes to fix the elastic-agent upgrade on MacOS and Windows.

## Why is it important?

This is the final block of code necessary to implement #2579 on Linux, MacOS and Windows. It will make use of the mapping described in the package manifest to correctly place agent files in directories containing the agent version and the hash in the name.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)
- ~~[ ] I have added an integration test or an E2E test~~

## Author's Checklist

- [ ]

## How to test this PR locally

To test this PR we can create multiple packages from the same agent commit and test if we can upgrade between them. The procedure is the same as described in #4174 but without the OS limitation and the known issues.

This change is supposed to be backward compatible so it can be tested also using older agent versions (like we do in our integration tests)

## Related issues

- Relates #2579 
- Relates #3950 
- Requires #4173
- Requires #4174
- Requires #4176 
- Requires #4177 

## Questions to ask yourself

- How are we going to support this in production? 
- How are we going to measure its adoption? 
- How are we going to debug this?
- What are the metrics I should take care of?
- ...

### CODE REVIEW

File: docs/upgrades.md [hunk 1]
```diff
@@ -61,3 +61,68 @@ sequenceDiagram
        end
     end
 ```
+
+### Introducing package manifest
+
+Starting from version 8.13.0 an additional file `manifest.yaml` is present in elastic-agent packages.
```
Comment (cmacknz) [inline docs/upgrades.md:67]: Documentation is great, let's add why we did this. To allow releasing versions of the agent package where only the component versions change, with the agent commit unchanged.

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 2]
```diff
@@ -258,30 +265,51 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
...
-		rollbackInstall(ctx, u.log, paths.Top(), hashedDir)
+		rollbackInstall(ctx, u.log, paths.Top(), hashedDir, currentVersionedHome)
 		return nil, err
 	}
 
-	if err := InvokeWatcher(u.log); err != nil {
+	minParsedVersionForNewUpdateMarker := agtversion.NewParsedSemVer(8, 13, 0, "", "")
+	var watcherExecutable string
+	if parsedVersion.Less(*minParsedVersionForNewUpdateMarker) {
+		// use the current agent executable for watch
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/upgrade.go:304]:
```suggestion
		// use the current agent executable for watch, if downgrading the old agent doesn't understand the current agent's path structure.
```

I think this could be a little bit clearer.

File: docs/upgrades.md (hunk 1 above)
Comment (pchila) [reply docs/upgrades.md:67]: Have a look at b87a013af7e3719dde052c5375eba0d92d95f26c
I tried adding a bit more context on that

Comment (cmacknz) [reply docs/upgrades.md:67]: Perfect, thanks!

File: dev-tools/packaging/templates/darwin/elastic-agent.tmpl [hunk 3]
```diff
@@ -6,6 +6,9 @@ set -e
 symlink="/Library/Elastic/Agent/elastic-agent"
 
 if test -L "$symlink"; then
-    ln -sfn "data/elastic-agent-{{ commit_short }}/elastic-agent.app/Contents/MacOS/elastic-agent" "$symlink"
+    symlinkTarget="data/elastic-agent-{{ commit_short }}/elastic-agent.app/Contents/MacOS/elastic-agent"
+    if test -f "data/elastic-agent-{{ agent_package_version }}{{ snapshot_suffix }}-{{ commit_short }}/elastic-agent.app/Contents/MacOS/elastic-agent"; then
```
Comment (cmacknz) [inline dev-tools/packaging/templates/darwin/elastic-agent.tmpl:10]: Out of curiosity why is this necessary? I though the package structure was static at the point of packaging?

File: internal/pkg/agent/application/upgrade/upgrade.go [hunk 4]
```diff
@@ -244,30 +251,51 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
...
+	current := agentInstall{
+		version:       version,
+		hash:          unpackRes.Hash,
+		versionedHome: unpackRes.VersionedHome,
+	}
+
+	previous := agentInstall{
+		version:       release.VersionWithSnapshot(),
+		hash:          release.Commit(),
+		versionedHome: currentVersionedHome,
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/upgrade.go:274]: It looks like you have swapped the `current` and `previous` names, since the `previous` version is getting the current values of version, hash, and version home.

File: dev-tools/packaging/templates/darwin/elastic-agent.tmpl (hunk 3 above)
Comment (pchila) [reply dev-tools/packaging/templates/darwin/elastic-agent.tmpl:10]: This is the script that fixes the links to the elastic-agent executable within the `elastic.app` directory. Now the path has potentially changed depending on how the package has been installed so the symlink target needs to be tested and set to the existing path

Comment (cmacknz) [reply dev-tools/packaging/templates/darwin/elastic-agent.tmpl:10]: Ah I missed that this script was executed at runtime, thanks. Thought it was part of the packaging process itself.

File: internal/pkg/agent/application/upgrade/upgrade.go (hunk 4 above)
Comment (pchila) [reply internal/pkg/agent/application/upgrade/upgrade.go:274]: It is correct: we have already rotated the symlink, so in the update we write the `current ` as the version we just upgraded to and the `previous` is the old agent installation. Confusing, I know but the nomenclature of the update marker is based on these definitions

Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/upgrade.go:274]: Thanks, this would be worth a comment in the code. The `previous.versionedHome.currentVersionedHome` seems suspicious when you read it.

Comment (pchila) [reply internal/pkg/agent/application/upgrade/upgrade.go:278]: 22606d7985ef6190e9ab0f89a60305b3136ffe13

File: internal/pkg/agent/application/upgrade/rollback.go [hunk 5]
```diff
@@ -46,7 +46,7 @@ func Rollback(ctx context.Context, log *logger.Logger, c client.Client, topDirPa
 		symlinkTarget = paths.BinaryPath(filepath.Join(paths.DataFrom(topDirPath), hashedDir), agentName)
 	}
 	// change symlink
-	if err := changeSymlinkInternal(log, topDirPath, symlinkPath, symlinkTarget); err != nil {
+	if err := ChangeSymlink(log, topDirPath, symlinkPath, symlinkTarget); err != nil {
```
Comment (ycombinator) [inline internal/pkg/agent/application/upgrade/rollback.go:49]: I think this could be left as-is or even just renamed to `changeSymlink`.  IOW, it doesn't look like this function is called from outside the `upgrade` package so no need to export it.

File: internal/pkg/agent/application/upgrade/step_download.go [hunk 6]
```diff
@@ -127,7 +122,7 @@ func (u *Upgrader) downloadArtifact(ctx context.Context, version, sourceURI stri
 	return path, nil
 }
 
-func (u *Upgrader) appendFallbackPGP(targetVersion string, pgpBytes []string) []string {
+func (u *Upgrader) appendFallbackPGP(tpv *agtversion.ParsedSemVer, pgpBytes []string) []string {
```
Comment (ycombinator) [inline internal/pkg/agent/application/upgrade/step_download.go:125]: Could we just continue to call it `targetVersion` instead of the less descriptive `tpv` please? 🙂

**Review by ycombinator — APPROVED** [review]
Left a couple of minor comments about naming; otherwise, LGTM!

**Comment by blakerouse** [comment]
The description says this is the final block of code required to implement #2579, but doesn't mark the issue as closed when this lands. Is a follow up needed? If so does that follow-up include an integration test? I would really like to see an integration tests that possibly takes the currently extracted Elastic Agent, re-archives it with a different manifest and performs an upgrage.

**Review by cmacknz — APPROVED** [review]
Same, a manual test of an upgrade where the agent commit hash is the same but the version is different succeeds. Thanks!

```
❯ sudo elastic-agent version
Binary: 8.16.1-SNAPSHOT (build: f6a3b8203762d074ae57eb8e3b28b15ca68cdea7 at 2024-02-05 21:06:09 +0000 UTC)
Daemon: 8.16.1-SNAPSHOT (build: f6a3b8203762d074ae57eb8e3b28b15ca68cdea7 at 2024-02-05 21:06:09 +0000 UTC)
~/Downloads ················································································· 04:23:55 PM
❯ sudo elastic-agent upgrade 8.16.2-SNAPSHOT --skip-verify --source-uri file:///Users/cmackenzie/go/src/github.com/elastic/elastic-agent/build/distributions
Upgrade triggered to version 8.16.2-SNAPSHOT, Elastic Agent is currently restarting
~/Downloads ············································································· 17s 04:24:18 PM
❯ sudo elastic-agent version
Binary: 8.16.2-SNAPSHOT (build: f6a3b8203762d074ae57eb8e3b28b15ca68cdea7 at 2024-02-05 21:11:59 +0000 UTC)
Daemon: 8.16.2-SNAPSHOT (build: f6a3b8203762d074ae57eb8e3b28b15ca68cdea7 at 2024-02-05 21:11:59 +0000 UTC)
```

Grabbing diagnostics the versions appear as expected as well:

```
 tree logs
logs
├── elastic-agent-8.16.1-SNAPSHOT-f6a3b8
│   ├── elastic-agent-20240205.ndjson
│   ├── elastic-agent-watcher-20240205-1.ndjson
│   └── elastic-agent-watcher-20240205.ndjson
└── elastic-agent-8.16.2-SNAPSHOT-f6a3b8
    ├── elastic-agent-20240205.ndjson
    ├── elastic-agent-watcher-20240205-1.ndjson
    └── elastic-agent-watcher-20240205.ndjson

2 directories, 6 files
```

Great job working through all this @pchila!

**Comment by cmacknz** [comment]
> The description says this is the final block of code required to implement https://github.com/elastic/elastic-agent/issues/2579, but doesn't mark the issue as closed when this lands. Is a follow up needed? If so does that follow-up include an integration test? I would really like to see an integration tests that possibly takes the currently extracted Elastic Agent, re-archives it with a different manifest and performs an upgrage.

Agree, I think I made the same comment in one of the earlier PRs. I did this exact test manually several times through these PRs, would be nice to automate it.

Let's track this in https://github.com/elastic/elastic-agent/issues/2780. I'll assign it to @pchila.

**Comment by pchila** [comment]
> The description says this is the final block of code required to implement #2579, but doesn't mark the issue as closed when this lands.

@blakerouse The big PR from the feature branch to main will close the issue. I would not consider the feature implemented if it's not available on main.

For the integration test we need to hack an existing package which is not something I am extremely fond of...

**Comment by blakerouse** [comment]
> > The description says this is the final block of code required to implement #2579, but doesn't mark the issue as closed when this lands.
> 
> @blakerouse The big PR from the feature branch to main will close the issue. I would not consider the feature implemented if it's not available on main.

Sorry was not paying attention, didn't realize this was for a feature branch.

> 
> For the integration test we need to hack an existing package which is not something I am extremely fond of...

If you have a better way then that would be great, otherwise I am not sure we can workaround it. Better to have a hacky test that tests a core feature, then no test at all.

**Review by blakerouse — APPROVED** [review]

File: internal/pkg/agent/application/upgrade/rollback.go (hunk 5 above)
Comment (pchila) [reply internal/pkg/agent/application/upgrade/rollback.go]: 280202026ca9bd5b2c78412f70ba5e77cd27f474

File: internal/pkg/agent/application/upgrade/step_download.go (hunk 6 above)
Comment (pchila) [reply internal/pkg/agent/application/upgrade/step_download.go]: b852a8371293bfcd808ddc63e0e1488e6c11bb09

**Comment by pchila** [comment]
>> For the integration test we need to hack an existing package which is not something I am extremely fond of...
> 
> If you have a better way then that would be great, otherwise I am not sure we can workaround it. Better to have a hacky test that tests a core feature, then no test at all.

@blakerouse Ideally we should package the commit twice with different versions and have both packages available on the runner but the current integration test framework does not allow that AFAIK.

I will have a look at what is possible in #2780 without sinking more time in the integration test framework

**Comment by pchila** [comment]
Integration test failed because of known flakiness in (linux-arm64-ubuntu-2204-upgrade)(sudo)  runner. Merging on feature branch an preparing the PR to main...
